language/tools/move-mv-llvm-compiler/src/stackless/module_context.rs

```diff
@@ -622,4 +626,216 @@ impl<'mm, 'up> ModuleContext<'mm, 'up> {
+        self.llvm_builder.call(ll_fn_deserialize, &input);
+
+        let insn_data = self.llvm_builder.getelementptr(
+            params,
+            &ll_sret.as_struct_type(),
+            0,
+            "instruction_data",
+        );
+        let insn_data_ptr = self.llvm_builder.getelementptr(
+            insn_data,
+            &str_slice_type.as_struct_type(),
+            0,
+            "insn_data_ptr",
+        );
+        let insn_data_ptr = self.llvm_builder.load(
+            insn_data_ptr,
+            self.llvm_cx.ptr_type(),
+            "insn_data_ptr_loaded",
+        );
+        let insn_data_len = self.llvm_builder.getelementptr(
+            insn_data,
+            &str_slice_type.as_struct_type(),
+            1,
+            "insn_data_len",
+        );
+        let insn_data_len = self.llvm_builder.load(
+            insn_data_len,
+            self.llvm_cx.int_type(64),
+            "insn_data_len_loaded",
+        );
+        let curr_bb = self.llvm_builder.get_insert_block();
+        let exit_bb = ll_fn_solana_entrypoint.insert_basic_block_after(curr_bb, "exit_bb");
+        for fun in entry_functions {
+            let entry = self.generate_global_str_slice(fun.llvm_symbol_name(&[]).as_str());
+
+            let func_name_ptr = self.llvm_builder.getelementptr(
+                entry.as_any_value(),
+                &str_slice_type.as_struct_type(),
+                0,
+                "entry_func_ptr",
+            );
+            let func_name_ptr = self.llvm_builder.load(
+                func_name_ptr,
+                self.llvm_cx.ptr_type(),
+                "entry_func_ptr_loaded",
+            );
+            let func_name_len = self.llvm_builder.getelementptr(
+                entry.as_any_value(),
+                &str_slice_type.as_struct_type(),
+                1,
+                "entry_func_len",
+            );
+            let func_name_len = self.llvm_builder.load(
+                func_name_len,
+                self.llvm_cx.int_type(64),
+                "entry_func_len_loaded",
+            );
+            let typarams = [insn_data_ptr, insn_data_len, func_name_ptr, func_name_len];
+            let condition = self.llvm_builder.call(ll_fn_str_cmp, &typarams);
+            let curr_bb = self.llvm_builder.get_insert_block();
+            let then_bb = ll_fn_solana_entrypoint.insert_basic_block_after(curr_bb, "then_bb");
+            let else_bb = ll_fn_solana_entrypoint.insert_basic_block_after(then_bb, "else_bb");
+            self.llvm_builder.build_cond_br(condition, then_bb, else_bb);
+            self.llvm_builder.position_at_end(then_bb);
+            let fn_name = fun.llvm_symbol_name(&[]);
+            let ll_fun = self.fn_decls.get(&fn_name).unwrap();
+            let ret = self.llvm_builder.call(*ll_fun, &[]);
+            self.llvm_builder.store(ret, retval);
+            self.llvm_builder.build_br(exit_bb);
+            self.llvm_builder.position_at_end(else_bb);
+        }
```

Comment:
I think we need one more "default" case emitted that calls `rt_abort` if the user passes an unrecognized entry point name. We do not want to fail silently.

Reply:
I wrote some quick tests that use `debug::print` and a `log` line to show different entry point callees actually being correctly called, along with the default/error case. The first calls `entry_point__foo`, the second `entry_point__bar` (those pass), and the third calls an unknown `entry_point__boo` (this should abort, but does not currently).
```
diff --git a/language/tools/move-mv-llvm-compiler/tests/rbpf-tests/entry-point01.json b/language/tools/move-mv-llvm-compiler/tests/rbpf-tests/entry-point01.json
new file mode 100644
index 000000000..44fae0b50
--- /dev/null
+++ b/language/tools/move-mv-llvm-compiler/tests/rbpf-tests/entry-point01.json
@@ -0,0 +1,14 @@
+{
+    "program_id": "DozgQiYtGbdyniV2T74xMdmjZJvYDzoRFFqw7UR5MwPK",
+    "accounts": [
+        {
+            "key": "524HMdYYBy6TAn4dK5vCcjiTmT2sxV6Xoue5EXrz22Ca",
+            "owner": "BPFLoaderUpgradeab1e11111111111111111111111",
+            "is_signer": false,
+            "is_writable": true,
+            "lamports": 1000,
+            "data": [0, 0, 0, 3]
+        }
+    ],
+    "instruction_data": [101, 110, 116, 114, 121, 95, 112, 111, 105, 110, 116, 95, 95, 102, 111, 111]
+}
diff --git a/language/tools/move-mv-llvm-compiler/tests/rbpf-tests/entry-point01.move b/language/tools/move-mv-llvm-compiler/tests/rbpf-tests/entry-point01.move
new file mode 100644
index 000000000..758e90495
--- /dev/null
+++ b/language/tools/move-mv-llvm-compiler/tests/rbpf-tests/entry-point01.move
@@ -0,0 +1,23 @@
+// input entry-point01.json
+// log 17
+
+module 0x10::debug {
+  native public fun print<T>(x: &T);
+}
+
+module 0xa000::entry_point {
+    use 0x10::debug;
+
+    public entry fun bar(): u64 {
+        let rv = 123;
+        debug::print(&rv);
+        rv
+    }
+
+    public entry fun foo(): u64 {
+        let rv = 17;
+        debug::print(&rv);
+        rv
+    }
+}
+
diff --git a/language/tools/move-mv-llvm-compiler/tests/rbpf-tests/entry-point02.json b/language/tools/move-mv-llvm-compiler/tests/rbpf-tests/entry-point02.json
new file mode 100644
index 000000000..985a0b65f
--- /dev/null
+++ b/language/tools/move-mv-llvm-compiler/tests/rbpf-tests/entry-point02.json
@@ -0,0 +1,14 @@
+{
+    "program_id": "DozgQiYtGbdyniV2T74xMdmjZJvYDzoRFFqw7UR5MwPK",
+    "accounts": [
+        {
+            "key": "524HMdYYBy6TAn4dK5vCcjiTmT2sxV6Xoue5EXrz22Ca",
+            "owner": "BPFLoaderUpgradeab1e11111111111111111111111",
+            "is_signer": false,
+            "is_writable": true,
+            "lamports": 1000,
+            "data": [0, 0, 0, 3]
+        }
+    ],
+    "instruction_data": [101, 110, 116, 114, 121, 95, 112, 111, 105, 110, 116, 95, 95, 98, 97, 114]
+}
diff --git a/language/tools/move-mv-llvm-compiler/tests/rbpf-tests/entry-point02.move b/language/tools/move-mv-llvm-compiler/tests/rbpf-tests/entry-point02.move
new file mode 100644
index 000000000..1609e657c
--- /dev/null
+++ b/language/tools/move-mv-llvm-compiler/tests/rbpf-tests/entry-point02.move
@@ -0,0 +1,23 @@
+// input entry-point02.json
+// log 123
+
+module 0x10::debug {
+  native public fun print<T>(x: &T);
+}
+
+module 0xa000::entry_point {
+    use 0x10::debug;
+
+    public entry fun bar(): u64 {
+        let rv = 123;
+        debug::print(&rv);
+        rv
+    }
+
+    public entry fun foo(): u64 {
+        let rv = 17;
+        debug::print(&rv);
+        rv
+    }
+}
+
diff --git a/language/tools/move-mv-llvm-compiler/tests/rbpf-tests/entry-point03.json b/language/tools/move-mv-llvm-compiler/tests/rbpf-tests/entry-point03.json
new file mode 100644
index 000000000..99105e65d
--- /dev/null
+++ b/language/tools/move-mv-llvm-compiler/tests/rbpf-tests/entry-point03.json
@@ -0,0 +1,14 @@
+{
+    "program_id": "DozgQiYtGbdyniV2T74xMdmjZJvYDzoRFFqw7UR5MwPK",
+    "accounts": [
+        {
+            "key": "524HMdYYBy6TAn4dK5vCcjiTmT2sxV6Xoue5EXrz22Ca",
+            "owner": "BPFLoaderUpgradeab1e11111111111111111111111",
+            "is_signer": false,
+            "is_writable": true,
+            "lamports": 1000,
+            "data": [0, 0, 0, 3]
+        }
+    ],
+    "instruction_data": [101, 110, 116, 114, 121, 95, 112, 111, 105, 110, 116, 95, 95, 98, 111, 111]
+}
diff --git a/language/tools/move-mv-llvm-compiler/tests/rbpf-tests/entry-point03.move b/language/tools/move-mv-llvm-compiler/tests/rbpf-tests/entry-point03.move
new file mode 100644
index 000000000..db731b7b3
--- /dev/null
+++ b/language/tools/move-mv-llvm-compiler/tests/rbpf-tests/entry-point03.move
@@ -0,0 +1,23 @@
+// input entry-point.json
+// abort 8192
+
+module 0x10::debug {
+  native public fun print<T>(x: &T);
+}
+
+module 0xa000::entry_point {
+    use 0x10::debug;
+
+    public entry fun bar(): u64 {
+        let rv = 123;
+        debug::print(&rv);
+        rv
+    }
+
+    public entry fun foo(): u64 {
+        let rv = 17;
+        debug::print(&rv);
+        rv
+    }
+}
+
```

Reply:
Correct. Currently it exits with exit code 0. I agree that it should abort, and I'll add a call to rt_abort.

Reply:
Could you add the test cases above to your patch?

Reply:
Will do.